resteasy-core/src/main/java/org/jboss/resteasy/core/providerfactory/ResteasyProviderFactoryImpl.java

```diff
@@ -1021,17 +1021,16 @@ public <T extends Throwable> ExceptionMapper<T> getExceptionMapperForClass(Class
     //   @Override
     public <T> AsyncResponseProvider<T> getAsyncResponseProvider(Class<T> type) {
         Class asyncType = type;
-        AsyncResponseProvider<T> mapper = null;
-        while (mapper == null) {
-            if (asyncType == null)
-                break;
-            Map<Class<?>, AsyncResponseProvider> asyncResponseProviders = getAsyncResponseProviders();
-            if (asyncResponseProviders != null)
-                mapper = asyncResponseProviders.get(asyncType);
-            if (mapper == null)
-                asyncType = asyncType.getSuperclass();
+
+        final Map<Class<?>, AsyncResponseProvider> asyncResponseProviders = getAsyncResponseProviders();
+        if (asyncResponseProviders == null)
+            return null;
+        for (Class<?> aClass : asyncResponseProviders.keySet()) {
+            if (aClass.isAssignableFrom(asyncType)) {
+                return asyncResponseProviders.get(aClass);
+            }
         }
-        return mapper;
+        return null;
     }
 
     public <T> AsyncClientResponseProvider<T> getAsyncClientResponseProvider(Class<T> type) {
```

Comment:
We should probably do the same thing from above here. Possibly with the `getAsyncStreamProvider()` too.